packages/core/src/components/icon/icon.tsx

```diff
@@ -101,7 +101,7 @@ export interface DefaultIconProps extends IntentProps, Props, DefaultSVGIconProp
  * @see https://stackoverflow.com/a/73795494/7406866
  */
 export interface IconComponent extends React.FC<IconProps<Element>> {
-    <T extends Element = Element>(props: IconProps<T>): React.ReactElement | null;
```

Comment:
Attempted to cherry-pick this type change into develop, but unfortunately encountered errors. This type fix is valid in React 18, but fails with React 16 types.

See type failures on branch based off of develop: https://github.com/palantir/blueprint/compare/gdouglas/test-Icon-compile-break

Reply:
I know I suggested the `React.ReactElement` to `React.ReactNode` change, but I'm realizing we might need something more dynamic to maintain both React 16 and 18 compatibility.

We could try this:

```ts
<T extends Element = Element>(props: IconProps<T>): ReturnType<React.FC< IconProps<Element>>> | null;
```

I think that should return `React.ReactNode` on `@types/react@18` and `React.ReactElement` on `@types/react@16`?

Reply:
I think this will work! Thanks for the suggestion. #7177